numpy/random/generator.pyx

```diff
@@ -3905,6 +3906,8 @@ cdef class Generator:
 
         arr = np.asarray(x)
 
+        axis = normalize_axis_index(axis, arr.ndim)
+
         # shuffle has fast-path for 1-d
         if arr.ndim == 1:
             # Return a copy if same memory
```

Comment:
Should axis!=0 raise ValueError here?

Reply:
`normalize_axis_index` should have already done the right thing by this line:

```python console
>>> import numpy
>>> numpy.core.multiarray.normalize_axis_index(0, 1)
0
>>> numpy.core.multiarray.normalize_axis_index(1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
numpy.AxisError: axis 1 is out of bounds for array of dimension 1
>>> numpy.core.multiarray.normalize_axis_index(-1, 1)
0
>>> numpy.core.multiarray.normalize_axis_index(-2, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
numpy.AxisError: axis -2 is out of bounds for array of dimension 1
>>> issubclass(numpy.AxisError, ValueError)
True
```

Reply:
Ok, great. Having seen both together permutation looks a lot less controversial than shuffle since the first thing permutation does is call asarray.

Reply:
Should there be tests of exceptions like https://github.com/numpy/numpy/blob/3deeea9522f5c512d52b3fe6d418d99b338eaece/numpy/random/tests/test_generator_mt19937.py#L598-L615